src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/WebSocketsTransport.cs

```diff
@@ -390,18 +393,45 @@ private async Task ProcessSocketAsync(WebSocket socket, Uri url, bool isReconnec
                 {
                     _application.Output.CancelPendingFlush();
                 }
+
+                // Should not throw
+                await receiving.ConfigureAwait(false);
             }
         }
 
-        if (_useStatefulReconnect && !_gracefulClose)
+        var cleanup = true;
+        try
         {
-            if (!UpdateConnectionPair())
+            if (_useStatefulReconnect && !_gracefulClose)
             {
-                return;
-            }
+                if (!UpdateConnectionPair())
+                {
+                    return;
+                }
 
-            await StartAsync(url, _webSocketMessageType == WebSocketMessageType.Binary ? TransferFormat.Binary : TransferFormat.Text,
-                cancellationToken: default).ConfigureAwait(false);
+                try
+                {
+                    await StartAsync(url, _webSocketMessageType == WebSocketMessageType.Binary ? TransferFormat.Binary : TransferFormat.Text,
+                        cancellationToken: default).ConfigureAwait(false);
+                    cleanup = false;
+                }
+                catch
+                {
+                    throw new InvalidOperationException("Reconnect attempt failed.");
+                }
+            }
+        }
+        finally
+        {
+            if (cleanup)
+            {
+                // Pipes will usually already be completed.
+                // If stateful reconnect fails we want to make sure the Pipes are cleaned up.
+                // And in rare cases where the websocket is closing at the same time StopAsync is called
```

Comment:
This seems better than before, but still susceptible to race conditions. In the scenario where `StopAsync` is called during a reconnect and `StartAsync` fails, the `Running` task never gets replaced which means that `StopAsync` can incorrectly return before the we fully complete the pipes.

Can we file an issue to make `Running` a TCS or something that only completes after we've fully stopped the transport?

Also, why do `StartReceiving` and `StartSending` start with "Start"?

Reply:
If `StartAsync` fails then `Running` hasn't been replaced yet so it's still pointing to the task that's waiting for this active `ProcessSocketAsync` to complete.